context.go

```diff
@@ -1,17 +1,35 @@
 package sync
 
-import "context"
+import (
+	"context"
+)
 
-type executorKey struct{}
+type executorKey struct {
+	name string
+}
+
+// HasContextExecutor returns true when the named executor is available in the context
+func HasContextExecutor(ctx context.Context, name string) bool {
+	return ctx.Value(executorKey{name: name}) != nil
+}
 
-func ContextExecutor(ctx context.Context) Executor {
-	executor, ok := ctx.Value(executorKey{}).(Executor)
-	if !ok {
-		return sequentialExecutor{}
+// ContextExecutor returns an executor in context with the given name, or a serial executor if none exists
+// and replaces the context with one that contains a new executor which won't deadlock if the context
```

Comment:
comment seems important but is not a complete thought (may be worth elaborating on the second point that got cut off)